nixos/modules/services/misc/xmrig.nix

```diff
@@ -51,14 +51,15 @@ with lib;
   };
 
   config = mkIf cfg.enable {
+    boot.kernelModules = [ "msr" ];
+
     systemd.services.xmrig = {
       wantedBy = [ "multi-user.target" ];
       after = [ "network.target" ];
       description = "XMRig Mining Software Service";
       serviceConfig = {
         ExecStartPre = "${cfg.package}/bin/xmrig --config=${configFile} --dry-run";
         ExecStart = "${cfg.package}/bin/xmrig --config=${configFile}";
-        DynamicUser = true;
```

Comment:
Can we add some capabilities instead of running as `root`?

Reply:
I tried the code below and it didn't work properly. I used [this link as a reference](https://man7.org/linux/man-pages/man7/capabilities.7.html)

```nix
systemd.services.xmrig = {
  wantedBy = [ "multi-user.target" ];
  after = [ "network.target" ];
  description = "XMRig Mining Software Service";
  serviceConfig = {
    ExecStartPre = "${cfg.package}/bin/xmrig --config=${configFile} --dry-run";
    ExecStart = "${cfg.package}/bin/xmrig --config=${configFile}";
    DynamicUser = true;
    AmbientCapabilities = "CAP_SYS_RAWIO";
  };
};
 ```

Reply:
IMO, if someone wants to harden its system NixOS offer tools for it already (you can always add something like `systemd.services.xmrig.serviceConfig.DynamicUser = true;` directly to your config).

Unless we have some kinda of policy for service hardening, I am fine with this change.

Reply:
> Unless we have some kinda of policy for service hardening

@thiagokokada shhh or else @mweinelt might hear you :eyes:

:smile:

Reply:
Given this is what upstream suggests (IIRC, from what I read on their site) I'm also fine with if we can't figure out how to make it work with `DynamicUser`. Bonus points if you apply any other potential hardening, though :+1:

Reply:
> @thiagokokada shhh or else @mweinelt might hear you eyes

I am not against hardening per see, but I think other systemd options are more effective than `DynamicUsers`, for example:

```
serviceConfig.PrivateTmp = "yes";
serviceConfig.ProtectHome = "yes";       
serviceConfig.ProtectSystem = "strict";
serviceConfig.NoNewPrivileges = "yes";
```

Etc etc.

Reply:
I added the options @thiagokokada provided to the service and it worked as expected!

Reply:
Maybe we can use `https://github.com/xmrig/xmrig/blob/dev/scripts/randomx_boost.sh` in `ExecStartPre` and make that privileged instead of the whole process?

It is mentioned here: https://xmrig.com/docs/miner/randomx-optimization-guide/msr#configuration

Reply:
According to the [documentation](https://www.freedesktop.org/software/systemd/man/systemd.service.html#id-1.8.3.6.2.4), something like this:

```
ExecStartPre = "!${pkgs.xmrig}/scripts/randomx_boost.sh"
``` 

However, this is assuming the scripts are in the package. Also, it is probably better to patch out the `modprobe msr` part.

Reply:
I feel that putting this script is not a good idea as it has some imperative commands in it. As @thiagokokada already mentioned I also think that the hardening might cause some problems later...